s3/src/test/scala/org/apache/pekko/stream/connectors/s3/scaladsl/S3IntegrationSpec.scala

```diff
@@ -48,30 +50,180 @@ trait S3IntegrationSpec
+      : Future[Unit] =
+    for {
+      bucketResponse <- S3.checkIfBucketExists(bucket)(implicitly, s3Attrs)
+      _ <- bucketResponse match {
+        case BucketAccess.AccessDenied =>
+          throw new RuntimeException(
+            s"Unable to create bucket: $bucket since it already exists however permissions are inadequate")
+        case BucketAccess.AccessGranted | BucketAccess.NotExists =>
+          for {
+            _ <- bucketResponse match {
+              case BucketAccess.AccessGranted =>
+                system.log.info(
+                  s"Deleting and recreating bucket: $bucket since it already exists with correct permissions")
+                TestUtils.cleanAndDeleteBucket(bucket, s3Attrs)
+              case _ =>
+                Future.successful(())
+            }
+            _ <- S3.makeBucket(bucket)(implicitly, s3Attrs)
+            // TODO: Figure out a way to properly test this with Minio, see https://github.com/akka/alpakka/issues/2750
+            _ <- if (versioning && this.isInstanceOf[AWSS3IntegrationSpec])
+              S3.putBucketVersioning(bucket, BucketVersioning().withStatus(BucketVersioningStatus.Enabled))(implicitly,
+                s3Attrs)
+            else
+              Future.successful(())
+            _ = bucketReference.set(bucket)
+          } yield ()
+      }
+    } yield ()
+
+  private val defaultBucketReference = new AtomicReference[String]()
+  def withDefaultBucket(testCode: String => Assertion): Assertion =
+    testCode(defaultBucketReference.get())
 
   // with dots forcing path style access
-  val bucketWithDots = "my.test.frankfurt"
+  private val bucketWithDotsReference = new AtomicReference[String]()
+  def withBucketWithDots(testCode: String => Assertion): Assertion =
+    testCode(bucketWithDotsReference.get())
+
+  private val nonExistingBucketReference = new AtomicReference[String]()
+  def withNonExistingBucket(testCode: String => Assertion): Assertion =
+    testCode(nonExistingBucketReference.get())
+
+  private val bucketWithVersioningReference = new AtomicReference[String]()
+  def withBucketWithVersioning(testCode: String => Assertion): Assertion =
+    testCode(bucketWithVersioningReference.get())
 
-  val bucketWithVersioning = "my-bucket-with-versioning"
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    val (bucketWithDots, defaultBucket, bucketWithVersioning, nonExistingBucket) = {
+      val baseGen = for {
+        bucketWithDots <- genBucketName(S3IntegrationSpec.BucketWithDots)
+        defaultBucket <- genBucketName(S3IntegrationSpec.DefaultBucket)
+        bucketWithVersioning <- genBucketName(S3IntegrationSpec.BucketWithVersioning)
+        nonExistingBucket <- genBucketName(S3IntegrationSpec.NonExistingBucket)
+      } yield (bucketWithDots, defaultBucket, bucketWithVersioning, nonExistingBucket)
+
+      if (randomlyGenerateBucketNames)
+        TestUtils.loopUntilGenRetrievesValue(baseGen.filter {
```

Comment:
The `TestUtils.loopUntilGenRetrievesValue` is necessary here because we are manually invoking the scalacheck generator rather than using the Scalatest property checking integration (due to the fact its occurring in an `afterAll` block rather than an actual test).

Due to this we and the fact that we are doing a `.filter` on the generator, we cannot use the standard `sample.get` because it will return `None` if the filter fails, hence why `TestUtils.loopUntilGenRetrievesValue` exists.